projects/button_adc/src/main.c

```diff
@@ -0,0 +1,59 @@
+#include "adc.h"
+#include "delay.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "interrupt.h"
+#include "log.h"
+#include "soft_timer.h"
+#include "wait.h"
+
+static void prv_adc_callback(const GpioAddress *address, void *context) {
+  AdcChannel *analogChannel = context;
+  uint16_t analogPinData = 0;
+  adc_read_raw(*analogChannel, &analogPinData);
+  LOG_DEBUG("Analog Pin data: %d\n", analogPinData);
+}
+
+int main(void) {
+  interrupt_init();
+  soft_timer_init();
+  gpio_init();
+  gpio_it_init();
+
+  GpioAddress analogPin = { .port = GPIO_PORT_A, .pin = 6 };
+  GpioAddress button = { .port = GPIO_PORT_B, .pin = 2 };
+
+  GpioSettings pinSettings = {
+    GPIO_DIR_IN,        //
+    GPIO_STATE_LOW,     //
+    GPIO_RES_NONE,      //
+    GPIO_ALTFN_ANALOG,  //
+  };
+  GpioSettings button_settings = {
+    .direction = GPIO_DIR_IN,  //
+    .state = GPIO_STATE_LOW,   //
+  };
+  InterruptSettings itSetting = {
+    INTERRUPT_TYPE_INTERRUPT,   //
+    INTERRUPT_PRIORITY_NORMAL,  //
+  };
+  gpio_init_pin(&analogPin, &pinSettings);
+  gpio_init_pin(&button, &button_settings);
+
+  adc_init(ADC_MODE_SINGLE);
+
+  AdcChannel analogChannel = NUM_ADC_CHANNELS;
+
+  adc_get_channel(analogPin, &analogChannel);
+  adc_set_channel(analogChannel, true);
+
+  LOG_DEBUG("ADC Channel: %d\n", analogChannel);
+
+  gpio_it_register_interrupt(&button, &itSetting, INTERRUPT_EDGE_FALLING, prv_adc_callback,
+                             &analogChannel);

```

Comment:
It's alright to pass around an address to a local variable (`analogChannel`, which should be in snake_case) in this case since it's declared in main, but in the future never ever pass around pointers to local variables if there's ever the slightest chance that the function declaring the local variable might return before the pointer is used! C may or may not reclaim the stack space your variable is stored in and you'll get super strange bugs. You can make that kind of static-state variable global and `static` to alleviate the issue.

Reply:
An thanks! Just wondering, why does making the variable `static` alleviate the issue of unreclaimed stack space (does C just handle memory of static variables better than local ones)? Then is setting `GpioAddress *analog_pin = context;` ok in the case of the callback function just because the function is `void` or should this be avoided too?

Reply:
It's not the `static` keyword per se that makes it safe to pass around pointers to static globals, it's that they're global - `static` just makes the variable not visible from other modules, which is good practice for module-local variables and functions. The issue with pointers to local variables is that when a function returns, the memory that its local variables are stored in is marked as overwritable, so if you have a pointer to a local variable of a function that has returned, when you dereference the pointer, C might have overwritten the variable. Global variables are stored in a different area of memory and their space is only reclaimed when the program exits, so they're safe to pass around pointers to.

It's okay in this case to pass around the pointer to `analog_pin` because it's declared in `main`, so you know that its space will never be reclaimed until `main` returns and the program exits. Setting `GpioAddress *analog_pin = context` is ok in the callback function, that's exactly what the context is for!